File: Editorial/1740_Find_Distance_in_a_Binary_Tree/Java/2_Lowest_Common_Ancestor_and_Breadth_First_Search.java
public class Solution {

    public int findDistance(TreeNode root, int p, int q) {
        TreeNode lca = findLCA(root, p, q);
        if (lca == null) {
            return -1;
        }

        Queue<TreeNode> bfs = new LinkedList<>();
        bfs.add(lca);
        int distance = 0, depth = 0;
        boolean foundp = false, foundq = false;

        while ((bfs.size() > 0) & ((foundp == false) | (foundq == false))) {
            int size = bfs.size();

            for (int i = 0; i < size; ++i) {
                TreeNode front = bfs.poll();
                if (front.val == p) {
                    distance += depth;
                    foundp = true;
                }
                if (front.val == q) {
                    distance += depth;
                    foundq = true;
                }
                if (front.left != null) {
                    bfs.add(front.left);
                }
                if (front.right != null) {
                    bfs.add(front.right);
                }
            }
            depth++;
        }

        return distance;
    }

    private TreeNode findLCA(TreeNode root, int p, int q) {
        if (root == null) {
            return null;
        }
        if ((root.val == p) | (root.val == q)) {
            return root;
        }

        TreeNode left = findLCA(root.left, p, q);
        TreeNode right = findLCA(root.right, p, q);

        if (left != null && right != null) {
            return root;
        }
        return (left != null) ? left : right;
    }
}

File: Editorial/1701-1800/1740_Find_Distance_in_a_Binary_Tree/Python3/2_Lowest_Common_Ancestor_and_Breadth_First_Search.py
class Solution:
    def findDistance(self, root, p, q):
        lca = self._find_LCA(root, p, q)
        bfs = deque([lca])
        distance = 0
        depth = 0
        foundp = False
        foundq = False
        while bfs and (not foundp or not foundq):
            size = len(bfs)
            for i in range(size):
                node = bfs.popleft()  # Dequeue the node
                if node.val == p:
                    distance += depth
                    foundp = True
                if node.val == q:
                    distance += depth
                    foundq = True
                if node.left:
                    bfs.append(node.left)  # Enqueue left child
                if node.right:
                    bfs.append(node.right)  # Enqueue right child
            depth += 1
        return distance

    def _find_LCA(self, root, p, q):
        if root is None or root.val == p or root.val == q:
            return root
        left = self._find_LCA(root.left, p, q)
        right = self._find_LCA(root.right, p, q)
        if left and right:
            return root
        return left if left else right
